packages/radio-group/src/vaadin-radio-group.js

```diff
@@ -73,510 +85,380 @@ class RadioGroupElement extends ThemableMixin(DirMixin(PolymerElement)) {
-      const newCheckedButton = this._radioButtons.filter((button) => button.value == newV)[0];
+    if (this.__selectedRadioButton && this.__selectedRadioButton.value === newValue) {
+      return;
+    }
+
+    const newSelectedRadioButton = this.__radioButtons.find((radioButton) => {
+      return radioButton.value === newValue;
+    });
 
-      if (newCheckedButton) {
-        this._selectButton(newCheckedButton);
-        this.setAttribute('has-value', '');
-      } else {
-        console.warn(`No <vaadin-radio-button> with value ${newV} found.`);
-      }
+    if (newSelectedRadioButton) {
+      this.__selectRadioButton(newSelectedRadioButton);
+      this.toggleAttribute('has-value', true);
+    } else {
+      console.warn(`The radio button with the value "${newValue}" was not found.`);
     }
   }
 
   /**
-   * Returns true if `value` is valid.
+   * Whenever `readonly` property changes on the group element,
+   * the method updates the `disabled` property for the radio buttons.
    *
-   * @return {boolean} True if the value is valid.
+   * @param {!boolean} newValue
+   * @param {!boolean} oldValue
+   * @private
    */
-  validate() {
-    return !(this.invalid = !this.checkValidity());
+  __readonlyChanged(newValue, oldValue) {
+    if (newValue !== oldValue) {
+      this.__updateRadioButtonsDisabledProperty();
+    }
   }
 
   /**
-   * Returns true if the current input value satisfies all constraints (if any)
-   * @return {boolean}
+   * Extends the method from `DisabledMixin`
+   * to update the `disabled` property for the radio buttons
+   * whenever the property changes on the group element.
+   *
+   * @param {!boolean} disabled
+   * @protected
    */
-  checkValidity() {
-    return !this.required || !!this.value;
+  _disabledChanged(disabled) {
+    super._disabledChanged(disabled);
+
+    this.__updateRadioButtonsDisabledProperty();
   }
 
   /**
-   * @param {number} idx
+   * Extends the method from `FocusMixin`
+   * to prevent removing the `focused` attribute
+   * when focus moves between radio buttons inside the group.
+   *
+   * @param {!FocusEvent} event
+   * @return {!boolean}
    * @protected
    */
-  _setFocusable(idx) {
-    this._radioButtons.forEach((radioButton, radioButtonIdx) => {
-      if (radioButtonIdx === idx) {
-        radioButton.focusElement.removeAttribute('tabindex');
-      } else {
-        radioButton.focusElement.setAttribute('tabindex', '-1');
-      }
-    });
-  }
-
-  /** @private */
-  _labelChanged(label) {
-    this._setOrToggleAttribute('has-label', !!label);
-  }
-
-  /** @private */
-  _errorMessageChanged(errorMessage) {
-    this._setOrToggleAttribute('has-error-message', !!errorMessage);
+  _shouldRemoveFocus(event) {
+    return !this.contains(event.relatedTarget);
   }
 
-  /** @private */
-  _helperTextChanged(helperText) {
-    this._setOrToggleAttribute('has-helper', !!helperText);
-  }
+  /**
+   * Extends the method from `FocusMixin`
```

Comment:
Let's use the same writing as in other components:

```suggestion
   * Override method inherited from `FocusMixin`
```

Reply:
v